_posts/2017/2017-03-10-creating-a-sales-analytics-app-with-ruby-on-rails-2.md

```diff
@@ -0,0 +1,84 @@
+The problem with this is that the library we were using for user management, the easy and simple `Devise`, didn't support that.
+Changing to username for the log in process was the first thing it came into our minds.
+
+This is probably the easiest solution but also, one that changed how the app worked.
+Now the merchants needed to think about a username and then, remember it.
+Our goal was to make the app the most accessible as possible and adding extra fields to remember was not the way to do it.
+
+The other answer that came to my mind was *just* to adapt the log in system to use encrypted emails.
+At that moment we didn't fully realize what was the scope of that task we were creating, the easy and simple Devise was about to become a pain all over the app in ways that we couldn't even dream of figuring it out.
+
+I started creating two methods for auto-encrypt and auto-decrypt any database field via Ruby Magic (AKA Metaprogramming).
+This way we could add encryption easily where needed. But then Devise said **no**.
+As it makes sense on an email field, the library turns that field into lowercase and as we where encrypting the email field and overwriting it, it turned `randomEncryptedStringForEmail` into `randomencryptedstringforemail`, which is not the same when decrypting it.
+
+After some hours of not understanding anything, I realized the lowercase issue and opted to create another field on the database for the merchant to store the encrypted value, so Devise couldn't lowercase it.
+With that, we had to change it everywhere, backend code and UI code, but to be honest, with `find and replace` happened to be an easy task.
+
+It wasn't until later when we saw the beast we created.
+The email field was used for **everything** (Both in Devise and in our App) that had to do with the User.
+Not only to update the email, update the password or restore password, but all over the core.
+As this kind of platforms usually have the User model in the middle, because you need to display all the data regarding to one specific user, we basically rewrote the all or at least most of the sessions and password code.
+That affected also to other parts of the Devise library as `reset_tokens` or `last_update` dates, which required even more core rewriting.
+
+## Hosting the app
+
+Now that we have dealt with all the issues on the app, here comes the hosting debate.
+Our team has experience on both [Heroku](https://www.heroku.com) and [AWS](https://aws.amazon.com) deployment so those are the main contenders.
+
+Following the philosophy on fast delivery, we started with Heroku, which allows us to deploy on one console command and to add some useful plugins like [Papertrail](https://papertrailapp.com) to store and search the logfiles or [New Relic](https://newrelic.com/) to check the application performance, something really critical to decide the host.
```

Comment:
Full stop after "performance"
Next sentence: So this was something really critical to decide on the host.